Add some documentation about the webhook (#2158)
elastic/cloud-on-k8s: docs/webhook.asciidoc

[float]
=== Architecture
The webhook is composed of 4 main components. Here is a brief description of each of them to understand how they interact, their naming, and how they are managed.

. A `ValidatingWebhookConfiguration` object that defines the validating webhook, named `elastic-es-validation.k8s.elastic.co`. It must be created before starting the operator. The `caBundle` field can be automatically managed as part of the automatic certificate management _(see below)_.
. A Kubernetes Service is used to expose the validating server, named `elastic-webhook-server`. It is in the same Namespace as the webhook server.
. A webhook server that actually validates the submitted resources. In ECK it is the operator itself when it is configured with the `webhook` role. See <<{p}-operator-config,Configuring ECK>> for more information about the `operator-roles` flag.
. A Secret containing the required certificates to secure the connection between the API server and the webhook server.
Like the ValidatingWebhookConfiguration, it must be created before starting the operator, even if it is empty. By default its name is `elastic-webhook-server-cert`.

> The content of this Secret and the lifecycle of the certificates are automatically managed for you. You can disable this feature if you want to setup the certificates by yourself or with https://github.com/jetstack/cert-manager[cert-manager]. See an example of the latter below.

Reviewer: ```suggestion
The content of this Secret and the lifecycle of the certificates are automatically managed for you. You can disable this feature if you want to manage the certificates yourself or with https://github.com/jetstack/cert-manager[cert-manager]. See an example of the latter below.
```

Reviewer: Also I don't remember if I brought this up before so sorry if I am repeating myself, but it may be worthwhile to be more explicit about what specifically we mean by managing the lifecycle.

We should also bring up here that it manages the CA bundle in the webhook config (which we reference above). All we have in the config reference:
https://github.com/elastic/cloud-on-k8s/blob/master/docs/operator-config.asciidoc
is that the flag
>Enables automatic webhook certificate management

so we should be clearer about the specifics either here or in that doc.

> The content of this Secret and the lifecycle of the certificates are automatically managed for you. ECK generates a dedicated and separate certificate authority and ensures that every components are rotated before the expiration date. The certificate authority is also used to configure the `caBundle` field of the `ValidatingWebhookConfiguration`. You can disable this feature if you want to manage the certificates yourself or with https://github.com/jetstack/cert-manager[cert-manager]. See an example of the latter below.

Reviewer: ```suggestion
The content of this Secret and the lifecycle of the certificates are automatically managed for you. ECK generates a dedicated and separate certificate authority and ensures that all components are rotated before the expiration date. The certificate authority is also used to configure the `caBundle` field of the `ValidatingWebhookConfiguration`. You can disable this feature if you want to manage the certificates yourself or with https://github.com/jetstack/cert-manager[cert-manager]. See an example of the latter below.
```

[float]
=== Managing the webhook certificate with cert-manager

> If ECK is currently running you first need to ensure that the automatic certificate management feature is disabled. This can be done by updating the operator deployment manifest and adding the `--manage-webhook-certs=false` flag.

Reviewer: ```suggestion
If ECK is currently running, you first must ensure that the automatic certificate management feature is disabled. This can be done by updating the operator deployment manifest and adding the `--manage-webhook-certs=false` flag.
```

Then, cert-manager v0.11+ must be installed as described in the https://docs.cert-manager.io/en/latest/getting-started/install/[cert-manager documentation].

The following example shows how to create all the resources that a webhook requires to function.

[source,yaml,subs="attributes,+macros"]
----
cat $$<<$$EOF | kubectl apply -f -
---
# this configures
# - a self signed cert-manager issuer
# - a service to point to the webhook
# - a self signed certificate for the webhook service
# - a validating webhook configuration

> # - updates the ECK operator statefulset to enable the webhook

Reviewer 2: this configures:
(...)
- the ECK operator statefulset to enable the webhook

              fieldPath: metadata.namespace
        - name: OPERATOR_IMAGE
          value: docker.elastic.co/eck/eck-operator:{eck_version}
        image: docker.elastic.co/eck/eck-operator:{eck_version}
        imagePullPolicy: Always
        name: manager
        ports:
        - containerPort: 9443
          name: webhook-server
          protocol: TCP
        volumeMounts:
        # this is the path controller-runtime looks for certs and should not be changed
        - mountPath: /tmp/k8s-webhook-server/serving-certs
          name: cert
          readOnly: true
        resources:
          limits:
            cpu: "1"
            memory: 150Mi
          requests:
            cpu: 100m
            memory: 50Mi
      volumes:
      - name: cert
        secret:
          defaultMode: 420
          secretName: elastic-webhook-server-cert
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: elastic-operator
      serviceAccountName: elastic-operator
      terminationGracePeriodSeconds: 10
  updateStrategy:
    rollingUpdate:
      partition: 0
    type: RollingUpdate

> EOF

Reviewer 2: :(
I hate that we have to duplicate and maintain up to date the operator manifest here. But I don't see an easy way out.

Reviewer: Agreed it is not great. When I first wrote this we did not have the main operator yaml configured for webhooks. Now we do though. So maybe we can just omit the sset entirely here and direct people to disable cert management in the operator? But keep the cert-manager resources

Author: It makes me realize that the sample is wrong here, we should include `--manage-webhook-certs=false` in the args of the container. I will remove the sset since it is clearly stated that "you first must ensure that the automatic certificate management feature is disabled"

[id="{p}-webhook"]
=== Validating webhook

> A validating webhook provides additional validation of Elasticsearch resources, it gives you an immediate feedback on the Elasticsearch manifests you submit and allow you to catch errors right away before ECK even tries to fulfill your request.

Reviewer: ```suggestion
A validating webhook provides additional validation of Elasticsearch resources: it provides immediate feedback on the Elasticsearch manifests you submit, allowing you to catch errors right away before ECK even tries to fulfill your request.
```

[float]
=== Architecture

> The webhook is composed of 4 main components, here is a brief description of each of them to understand how they interact,

Reviewer: ```suggestion
The webhook is composed of 4 main components. Here is a brief description of each of them to understand how they interact, their naming, and how they are managed.
```

> their naming and how they are managed.

Reviewer: ```suggestion
```

> . A `ValidatingWebhookConfiguration` object that defines the validating webhook. In ECK it is named `elastic-es-validation.k8s.elastic.co`. It must be created before to start the operator, the `caBundle` field can be automatically managed as part of the

Reviewer: ```suggestion
. A `ValidatingWebhookConfiguration` object that defines the validating webhook, named `elastic-es-validation.k8s.elastic.co`. It must be created before starting the operator. The `caBundle` field can be automatically managed as part of the automatic certificate management _(see below)_.
```

It only needs to be created beforehand if using the automatic management, correct? That may be too nitpicky to be worth explaining though.

> automatic certificate management _(see below)_.

Reviewer: ```suggestion
```

Some line wrapping that doesn't seem necessary

> . A Kubernetes Service is used to expose the validating server, in ECK its name is fixed to `elastic-webhook-server`.

Reviewer: ```suggestion
. A Kubernetes Service is used to expose the validating server, named `elastic-webhook-server`. It is in the same Namespace as the webhook server.
```

> It lives in the same Namespace as the webhook server.

Reviewer: ```suggestion
```

> . A webhook server that actually validates the submitted resources. In ECK it is the operator itself when it

Reviewer: Same here with the wrapping

> is configured to have the `webhook` role. See <<{p}-operator-config,Configuring ECK>> if you want to learn more about the

Reviewer: ```suggestion
is configured with the `webhook` role. See <<{p}-operator-config,Configuring ECK>> if you want to learn more about the
```

`operator-roles` flag.

> . A Secret which contains the required certificates to secure the connection between the API server and the webhook.

Reviewer: ```suggestion
. A Secret containing the required certificates to secure the connection between the API server and the webhook server.
```

> Like the ValidatingWebhookConfiguration it must be created before starting the operator, even if it is empty.

Reviewer: ```suggestion
Like the ValidatingWebhookConfiguration, it must be created before starting the operator, even if it is empty.
```

and again here with the wrapping. We should only need it between paragraphs.

If ECK has been installed with the provided manifest its name is `elastic-webhook-server-cert`.
The content of this Secret and the lifecycle of the certificates are automatically managed for you. You can disable

> this feature if you want to setup the certificates by yourself or with the cert-manager, see an example for the latter below.

Reviewer: ```suggestion
this feature if you want to setup the certificates by yourself or with https://github.com/jetstack/cert-manager[cert-manager]. See an example of the latter below.
```

[float]

> === Managing webhook certificate with cert-manager

Reviewer: ```suggestion
=== Managing the webhook certificate with cert-manager
```

> If the operator is currently running you first need to ensure that the automatic certificate management is disabled.

Reviewer: ```suggestion
If ECK is currently running you first need to ensure that the automatic certificate management feature is disabled.
```

This can be done by updating the operator deployment manifest and adding the `--manage-webhook-certs=false` flag.

> Then, if it is not already done you have to install cert-manager v0.11+ as described in the https://docs.cert-manager.io/en/latest/getting-started/install/[cert-manager documentation].

Reviewer: ```suggestion
Then, cert-manager v0.11+ must be installed as described in the https://docs.cert-manager.io/en/latest/getting-started/install/[cert-manager documentation].
```

> Below is a complete example to create all of the resources necessary for the validating webhook to function.

Reviewer 3: The following example shows how to create all the resources that a webhook requires to function.

[id="{p}-webhook"]
== Validating webhook

> A validating webhook provides additional validation of Elasticsearch resources beyond what can be specified in OpenAPI specs.

Reviewer: I agree, do you think it would be useful to just strike everything from "beyond" and on?

This allows advanced validation to help prevent Elasticsearch resources that are invalid and will not function correctly

> from being admitted by the Kubernetes API server, which can aid troubleshooting.

Reviewer 4: I would go maybe for a bit more general introduction something along the lines of: validating webhooks give you immediate feedback on the Elasticsearch manifests you submit and allow you to catch errors right away before ECK even tries to fullfill your request.

[float]
=== Architecture
The webhook is composed of 4 main components, here is a brief description of each of them to understand how they interact,
their naming and how they are managed.

. A `ValidatingWebhookConfiguration` object that defines the validating webhook. In ECK it is named `elastic-es-validation.k8s.elastic.co`. It must be created before to start the operator, the `caBundle` field can be automatically managed as part of the
automatic certificate management _(see below)_.
. A Kubernetes Service is used to expose the validating server, in ECK its name is fixed to `elastic-webhook-server`.

> It lives in the same Namespace that the webhook server.

Reviewer 4: ...in the same Namespace **as** the...

. A webhook server that actually validates the submitted resources. In ECK it is the operator itself when it

> is configured to act the `webhook` role. See <<{p}-operator-config,Configuring ECK>> if want to learn more about the

Reviewer 4: ...when it is configured to **have** the `webhook` role...
... if **you** want to learn...

`operator-roles` flag.
. A Secret which contains the required certificates to secure the connection between the API server and the webhook.

> Like the ValidatingWebhookConfiguration it must be created before to start the operator, even if it is empty.

Reviewer 4: ...before **starting** the operator...

> If ECK has been installed with the provided manifest its name is `elastic-webhook-server-cert`.

Reviewer 4: > provided

Do you mean the "official" manifests from our downloads page or the one included on this documentation page? In either case I think we should change the sentence to make clear which manifests we are referring to.

Reviewer: Maybe just "By default, its name is..."?

> The content of this Secret and the lifecycle of the certificates are automatically handled for you. You can disable

Reviewer 4: ... automatically **managed** for you...

> this feature if you want to manage the certificates by yourself or with the cert-manager, see an example below.

Reviewer 4: ...if you want to **set up**... see an example **for the latter** below ... (not 💯 about this last one)

[float]
=== Managing webhook certificate with cert-manager

If the operator is currently running you first need to ensure that the automatic certificate management is disabled.
This can be done by updating the operator deployment manifest and adding the `--manage-webhook-certs=false` flag.

> Then if it is not already done you have to install cert-manager v0.11+ as described in their https://docs.cert-manager.io/en/latest/getting-started/install/[documentation].

Reviewer 4: Then, ... (I think, but commas are not my strength)

... described in the **cert-manager** documenation ...

----
cat $$<<$$EOF | kubectl apply -f -
---
# this configures
# - a self signed cert-manager issuer
# - a service to point to the webhook
# - a self signed certificate for the webhook service
# - a validating webhook configuration
# - updates the ECK operator statefulset to enable the webhook
apiVersion: cert-manager.io/v1alpha2
kind: Issuer
metadata:
  name: selfsigned-issuer
  namespace: elastic-system
spec:
  selfSigned: {}
---
apiVersion: cert-manager.io/v1alpha2
kind: Certificate
metadata:
  name: elastic-webhook
  namespace: elastic-system
spec:
  commonName: elastic-webhook.elastic-system.svc
  dnsNames:
  - elastic-webhook.elastic-system.svc.cluster.local
  - elastic-webhook.elastic-system.svc
  issuerRef:
    kind: Issuer
    name: selfsigned-issuer
  secretName: elastic-webhook-server-cert
---
apiVersion: v1
kind: Service
metadata:
  name: elastic-webhook-server
  namespace: elastic-system
spec:
  ports:
  - port: 443
    protocol: TCP
    targetPort: 9443
  selector:
    control-plane: elastic-operator
  sessionAffinity: None
  type: ClusterIP
---
apiVersion: admissionregistration.k8s.io/v1beta1
kind: ValidatingWebhookConfiguration
metadata:

>   name: elastic-es-validation.k8s.elastic.co

Reviewer 4: Should this not be `elastic-webhook.k8s.elastic.co`

        image: docker.elastic.co/eck/eck-operator:{eck_version}
        imagePullPolicy: Always
        name: manager
        ports:
        - containerPort: 9443
          name: webhook-server
          protocol: TCP
        volumeMounts:
        # this is the path controller-runtime looks for certs and should not be changed
        - mountPath: /tmp/k8s-webhook-server/serving-certs
          name: cert
          readOnly: true
        resources:
          limits:
            cpu: "1"
            memory: 150Mi
          requests:
            cpu: 100m
            memory: 50Mi
      volumes:
      - name: cert
        secret:
          defaultMode: 420
          secretName: elastic-webhook-server-cert
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: elastic-operator
      serviceAccountName: elastic-operator
      terminationGracePeriodSeconds: 10
  updateStrategy:
    rollingUpdate:
      partition: 0
    type: RollingUpdate
EOF
----

NOTE: This example assumes that you have installed the operator in the `elastic-system` namespace.

[float]
=== Troubleshooting

> Webhooks rely on a good connectivity between the Kubernetes API server and the operator.

Reviewer 4: "Webhooks require network connectivity between" is maybe marginally better, definitely remove the "a"

> See <<{p}-troubleshooting,Webhook troubleshooting>> for more information about some known problems with some Kubernetes providers.

Reviewer 4: Is this supposed to link directly to the webhook section on the troubleshooting page, because that is what I would expect

Author: Yes, I have checked on the preview and the link is fixed now.
